src/Lib.P2P/PeerConnection.cs

```diff
@@ -0,0 +1,441 @@
+                }
+            }
+
+            if (!enumerable.Any())
+            {
+                throw new Exception($"Protocol '{name}' is not registered.");
+            }
+            
+            throw new Exception($"{RemotePeer.Id} does not support protocol '{name}'.");
+        }
+
+        /// <summary>
+        ///   Starts reading messages from the remote peer.
+        /// </summary>
+        internal async Task ReadMessagesAsync(CancellationToken cancel)
+        {
+            _log.Debug($"start reading messsages from {RemoteAddress}");
+
+            // TODO: Only a subset of protocols are allowed until
+            // the remote is authenticated.
+            IPeerProtocol protocol = new Multistream1();
+            try
+            {
+                while (!cancel.IsCancellationRequested && Stream != null)
+                {
+                    await protocol.ProcessMessageAsync(this, Stream, cancel).ConfigureAwait(false);
+                }
+            }
+            catch (IOException e)
+            {
+                _log.Error("reading message failed " + e.Message);
+
+                // eat it.
+            }
+            catch (Exception e)
+            {
+                if (!cancel.IsCancellationRequested && Stream != null)
+                {
+                    _log.Error("reading message failed", e);
+                }
+            }
+
+            // Ignore any disposal exceptions.
+            try
+            {
+                Stream?.DisposeAsync();
+            }
+            catch (Exception)
+            {
+                // eat it.
+            }
+
+            _log.Debug($"stop reading messsages from {RemoteAddress}");
+        }
+
+        /// <summary>
+        ///   Starts reading messages from the remote peer on the specified stream.
+        /// </summary>
+        internal async Task ReadMessagesAsync(Stream stream, CancellationToken cancel)
+        {
+            IPeerProtocol protocol = new Multistream1();
+            try
+            {
+                while (!cancel.IsCancellationRequested && stream != null && stream.CanRead)
+                {
+                    await protocol.ProcessMessageAsync(this, stream, cancel).ConfigureAwait(false);
+                }
+            }
+            catch (EndOfStreamException)
+            {
+                // eat it.
+            }
+            catch (Exception e)
+            {
+                if (!cancel.IsCancellationRequested && stream != null)
+                {
+                    _log.Error($"reading message failed {RemoteAddress} {RemotePeer}", e);
+                }
+            }
+        }
+
+        #region IDisposable Support
+
+        private bool _disposedValue; // To detect redundant calls
+
+        /// <summary>
+        ///   Signals that the connection is closed (disposed).
+        /// </summary>
+        public event EventHandler<PeerConnection> Closed;
+
+        /// <summary>
+        ///  TODO
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Complete the task associated to this 'TODO' comment.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)